### PR TITLE
[JENKINS-62409] Use plugin pom 4.0

### PR DESCRIPTION
**Pull request by ikedam**
https://issues.jenkins-ci.org/browse/JENKINS-62409

* Use plugin pom 4.0: https://github.com/jenkinsci/plugin-pom/tree/plugin-4.2
* Use jenkins bom: https://github.com/jenkinsci/bom/tree/bom-10
* Specify CI configurations explicitly instead of `recommendedConfigurations`
    * https://github.com/jenkins-infra/pipeline-library/pull/145
    * I actually don't think that's a good idea to specify configurations explicitly as we have to update Jenkinsfile manually. But Jenkins-2.164.1, the current target with `recommendedConfigurations` is old and I believe the root cause of #119 is the issue of `recommendedConfigurations`. 

I already removed the dependency to matrix-auth in #122 .

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -26,13 +26,13 @@
     <properties>
         <revision>1.45</revision>
         <changelist>-SNAPSHOT</changelist>
-        <jenkins.version>2.121.1</jenkins.version>
+        <jenkins.version>2.164.3</jenkins.version>
```
Comment (ikedam) [inline pom.xml:29]: 2.164.1 is the oldest LTS available in parent-pom 4.x.
And 2.164.1 caused following test failure:
```
[ERROR] org.jvnet.hudson.test.PluginAutomaticTestBuilder$OtherTests.testPluginActive  Time elapsed: 0.037 s  <<< ERROR!
java.lang.Error: Plugin junit failed to start
        at org.jvnet.hudson.test.PluginAutomaticTestBuilder$OtherTests.testPluginActive(PluginAutomaticTestBuilder.java:99)
...
Caused by: java.io.IOException: JUnit Plugin version 1.29 failed to load.
 - You must update Jenkins from version 2.164.1 to version 2.164.3 or later to run this plugin.
        at hudson.PluginWrapper.resolvePluginDependencies(PluginWrapper.java:821)
...
```

junit-plugin 1.29 is come from maven-plugin 3.1.2, and upgrading maven-plugin can be an option.

File: pom.xml [hunk 2]
```diff
@@ -26,13 +26,13 @@
...
         <java.level>8</java.level>
-        <workflow-api-plugin.version>2.30</workflow-api-plugin.version>
         <useBeta>true</useBeta>
```
Comment (ikedam) [inline pom.xml:31]: `useBeta` is still required.
```
[ERROR] hudson/plugins/copyartifact/CopyArtifact:671 jenkins/util/VirtualFile.list(Ljava/lang/String;Ljava/lang/String;Z)Ljava/util/Collection; is still in beta
[ERROR] hudson/plugins/copyartifact/CopyArtifact:689 jenkins/util/VirtualFile.readLink()Ljava/lang/String; is still in beta
[ERROR] hudson/plugins/copyartifact/CopyArtifact:696 jenkins/util/VirtualFile.toExternalURL()Ljava/net/URL; is still in beta
[ERROR] hudson/plugins/copyartifact/CopyArtifact:725 jenkins/util/VirtualFile.mode()I is still in beta
```

File: pom.xml [hunk 3]
```diff
@@ -196,57 +176,36 @@
...
-        <version>${workflow-scm-step.version}</version>
         <scope>test</scope>
       </dependency>
       <dependency>
         <!-- for dependency of GitSampleRepoRule -->
-        <groupId>org.jenkins-ci.plugins.workflow</groupId>
-        <artifactId>workflow-scm-step</artifactId>
-        <version>${workflow-scm-step.version}</version>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>scm-api</artifactId>
```
Comment (ikedam) [inline pom.xml:196]: https://github.com/jenkinsci/copyartifact-plugin/pull/119/files#r439700373

**Comment by ikedam** [comment]
I'll invite reviewers after I see CI passed.

**Comment by ikedam** [comment]
The failure looks caused by a machine issue. Let’s retry.

File: src/test/java/hudson/plugins/copyartifact/ParameterizedBuildSelectorTest.java [hunk 4]
```diff
@@ -131,7 +132,9 @@ public void testWorkflow() throws Exception {
         ));
         
         VirtualFile vf = b.getArtifactManager().root().child("artifact.txt");
-        assertEquals("foobar", IOUtils.toString(vf.open()));
+        try(InputStream in = vf.open()) {
+            assertEquals("foobar", IOUtils.toString(in, "UTF-8"));
+        }
```
Comment (ikedam) [inline src/test/java/hudson/plugins/copyartifact/ParameterizedBuildSelectorTest.java:137]: To release the file handle on Windows.
Windows denies deleting files when their handles are held.

File: Jenkinsfile [hunk 5]
```diff
@@ -1 +1,11 @@
-buildPlugin(configurations: buildPlugin.recommendedConfigurations())
+// Specify configurations explicitly to test against newer LTS.
+// See also https://github.com/jenkins-infra/pipeline-library/pull/145
+
+// This should be updated manually.
+def recentLts = '2.222.4'
+
+buildPlugin(configurations: [
+  [ platform: 'linux', jdk: '8', jenkins: null ],
```
Comment (jglick) [inline Jenkinsfile:8]:
```suggestion
  [ platform: 'linux', jdk: '8' ],
```

File: Jenkinsfile [hunk 6]
```diff
@@ -1 +1,11 @@
...
+  [ platform: 'windows', jdk: '8', jenkins: recentLts, javaLevel: '8' ],
+  [ platform: 'linux', jdk: '11', jenkins: recentLts, javaLevel: '8' ],
```
Comment (jglick) [inline Jenkinsfile:10]:
```suggestion
  [ platform: 'windows', jdk: '8', jenkins: recentLts ],
  [ platform: 'linux', jdk: '11', jenkins: recentLts ],
```

File: pom.xml (hunk 2 above)
Comment (jglick) [reply pom.xml:31]: Yes. I do plan to remove the beta mark from at least `list` and `toExternalURL`. I am less sure about `readLink` and `mode`, which are indeed used here though my opinion is that users would be better off not relying on some filesystem-specific features.

File: pom.xml [hunk 7]
```diff
@@ -26,13 +26,13 @@
...
         <java.level>8</java.level>
-        <workflow-api-plugin.version>2.30</workflow-api-plugin.version>
         <useBeta>true</useBeta>
-        <workflow-cps-plugin.version>2.32</workflow-cps-plugin.version>
-        <git-plugin.version>3.0.5</git-plugin.version>
-        <workflow-scm-step.version>2.4</workflow-scm-step.version>
+        <maven-plugin.version>3.1.2</maven-plugin.version>
+        <!-- versions for test dependencies -->
+        <compress-artifacts.version>1.10</compress-artifacts.version>
+        <jenkins-test-harness-tools.version>2.0</jenkins-test-harness-tools.version>
```
Comment (jglick) [inline pom.xml:35]: Could just inline. Generally I only recommend using a POM property for a dependency version in cases where two classifiers (e.g., main and `tests`) of the same artifact are needed.

File: pom.xml [hunk 8]
```diff
@@ -66,117 +66,97 @@
...
         <optional>true</optional>
       </dependency>
       <dependency>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>matrix-project</artifactId>
-        <version>1.7.1</version>
       </dependency>
       <dependency>
         <groupId>org.jenkins-ci.plugins.workflow</groupId>
         <artifactId>workflow-support</artifactId>
```
Comment (jglick) [inline pom.xml:92]: Probably could be deleted as a transitive dep.

File: pom.xml [hunk 9]
```diff
@@ -66,117 +66,97 @@
...
       </dependency>
       <dependency>
         <groupId>org.jenkins-ci.plugins.workflow</groupId>
         <artifactId>workflow-durable-task-step</artifactId>
-        <version>2.18</version>
         <scope>test</scope>
       </dependency>
       <dependency>
         <groupId>org.jenkins-ci.plugins.workflow</groupId>
         <artifactId>workflow-step-api</artifactId>
```
Comment (jglick) [inline pom.xml:124]: ditto

File: pom.xml [hunk 10]
```diff
@@ -66,117 +66,97 @@
...
-        <version>2.13</version>
         <scope>test</scope>
       </dependency>
       <dependency>
         <groupId>org.jenkins-ci.plugins.workflow</groupId>
         <artifactId>workflow-api</artifactId>
```
Comment (jglick) [inline pom.xml:129]: ditto

File: pom.xml (hunk 2 above)
Comment (ikedam) [reply pom.xml:31]: 👍 
I definitely agree that. I wrote that in the documentation: https://github.com/jenkinsci/copyartifact-plugin#recommended-usage-of-copyartifact
But copyartifact should keep backward compatibilities and continues to use `mode` and `readLink`...

**Pull request merged**